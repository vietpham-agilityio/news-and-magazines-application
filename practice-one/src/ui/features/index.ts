export * from '@/ui/features/TagGroup';
export * from '@/ui/features/SidebarCard';
export * from '@/ui/features/HashtagGroup';

export { default as BadgeGroup } from '@/ui/features/BadgeGroup';
export { default as CommentGroup } from '@/ui/features/CommentGroup';
export { default as CategoryMenu } from '@/ui/features/CategoryMenu';
export { default as AuthorInformation } from '@/ui/features/AuthorInformation';
export { default as CategoryBurgerMenu } from '@/ui/features/CategoryBurgerMenu';

// sections
export { default as PostsGrid } from '@/ui/sections/PostsGrid';
export { default as PostDetail } from '@/ui/sections/PostDetail';
export { default as RightSidebar } from '@/ui/sections/RightSidebar';
export { default as PostsByCategory } from '@/ui/sections/PostsByCategory';
export { default as PostsByAttribute } from '@/ui/sections/PostsByAttribute';