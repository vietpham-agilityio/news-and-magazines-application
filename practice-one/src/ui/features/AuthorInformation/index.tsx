// services
import { getAuthorById, getPostDataById } from '@/services';

// components
import { Avatar, Button, Information } from '@/ui/components';

// icons
import { Plus } from '@/ui/components/Icons';

interface IProps {
  postId: string;
}

const AuthorInformation = async ({ postId }: IProps) => {
  const postResponse = await getPostDataById(postId);
  const { authorId } = postResponse.data.attributes;

  const { data: authorResponse } = await getAuthorById(authorId);
  const { avatarUrl, name } = authorResponse.attributes;

  return (
    <div className="bg-white-90 flex p-[15px] mb-[25px] rounded-xl">
      <Avatar width={87} height={87} url={avatarUrl} alt="Avatar author" />
      <div className="flex-1 ml-2.5 mt-3.5">
        <Information name={name} information="27 posts" />
        <div className="mt-2">
          <Button
            label="Follow"
            leftIcon={<Plus />}
            bgColor="bg-gradient-to-bl from-sunSet-0 to-sunSet-100"
            textColor="text-white-100"
          />
        </div>
      </div>
    </div>
  );
};

export default AuthorInformation;
