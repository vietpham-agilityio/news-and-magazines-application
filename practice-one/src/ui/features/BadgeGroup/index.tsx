// services
import {
  getCategoryById,
  getCategoryByPostId,
  getCommentByPostId,
  getPostDataById,
} from '@/services';

// componens
import { Tag } from '@/ui/components';

// icons
import { ChatBox, Schedule, File } from '@/ui/components/Icons';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

export default async function BadgeGroup({ postId }: IProps) {
  const postResponse = await getPostDataById(postId);
  const { data: listComments } = await getCommentByPostId(postId);

  const { data: postCategoriesDataResponse } = await getCategoryByPostId({
    postId,
  });
  const categoryId =
    postCategoriesDataResponse[0]?.attributes.categoryId ?? '1';
  const categoryIdToNumber = parseInt(categoryId, 10);
  const { data: categoryDataResponse } =
    await getCategoryById(categoryIdToNumber);
  const { name: categoryType } = categoryDataResponse.attributes;

  const attributes = postResponse?.data?.attributes;
  const publicDate = formatDate(attributes?.publicationDate) ?? '';
  const commentCount = listComments?.length ?? '0';

  return (
    <div className="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4">
      <Tag icon={<Schedule />} content={publicDate} />
      <Tag
        icon={<ChatBox />}
        content={`Comments: ${commentCount}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<File />}
        content={`Categories: ${categoryType}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<ChatBox />}
        content={`${commentCount}`}
        additionalClass="sm:hidden"
      />
      <Tag
        icon={<File />}
        content={`${categoryType}`}
        additionalClass="sm:hidden"
      />
    </div>
  );
}
