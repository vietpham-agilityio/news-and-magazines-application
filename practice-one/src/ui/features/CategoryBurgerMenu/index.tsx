// services
import { getCategoryData } from '@/services';

// types
import { ICategoryItem, IMenu } from '@/types';

// componens
import { BurgerMenu } from '@/ui/components';

const CategoryBurgerMenu = async () => {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse?.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <BurgerMenu listCategories={listCategories} />;
};

export default CategoryBurgerMenu;
