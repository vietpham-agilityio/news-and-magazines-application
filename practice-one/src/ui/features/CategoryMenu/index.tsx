// services
import { getCategoryData } from '@/services';
import { ICategoryItem, IMenu } from '@/types';

// componens
import { Menu } from '@/ui/components';

export default async function CategoryMenu() {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <Menu name="Categories" listMenu={listCategories} />;
}
