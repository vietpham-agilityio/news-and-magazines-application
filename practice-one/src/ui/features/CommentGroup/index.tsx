// components
import { Comment, Typography } from '@/ui/components';

// services
import { getCommentByPostId } from '@/services';

// types
import { FontWeight } from '@/types';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

const CommentGroup = async ({ postId }: IProps) => {
  const { data: listComments } = await getCommentByPostId(postId);

  if (listComments.length === 0) {
    return (
      <div className="flex justify-center py-10">
        <Typography
          tag="h3"
          weight={FontWeight.Light}
          additionalClasses="text-sm text-black-75 leading-5"
        >
          This article has no comments
        </Typography>
      </div>
    );
  }

  return (
    <div className="flex flex-col gap-9 mb-10 sm:mb-12 lg:mb-15">
      {listComments.map(comment => {
        const { name, content, createdAt } = comment.attributes;
        return (
          <Comment
            key={comment.id}
            userName={name}
            content={content}
            publicDate={formatDate(createdAt)}
          />
        );
      })}
    </div>
  );
};

export default CommentGroup;
