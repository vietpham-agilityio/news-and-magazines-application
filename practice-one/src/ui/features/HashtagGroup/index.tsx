// types
import { IHashtag } from "@/types";
import { LazyImage } from "@/ui/components";

interface IProps {
  listHashtags: IHashtag[];
}

export const HashtagGroup = ({ listHashtags }: IProps) =>
  listHashtags.map((hashtag) => {
    const { id, name, imageUrl } = hashtag;
    return (
      <div key={id} className="mx-3 py-2.5 rounded-xl cursor-pointer">
        <LazyImage
         src={imageUrl}
         alt={name}
         width={170}
         height={48}
        />
      </div>
    );
  });
