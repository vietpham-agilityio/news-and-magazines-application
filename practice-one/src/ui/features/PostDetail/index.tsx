// services
import {
  getCategoryById,
  getCategoryByPostId,
  getCommentByPostId,
  getPostDataById,
} from '@/services';

// utils
import generateRGBDataURL from '@/utils/color';

// componens
import Image from 'next/image';
import { Typography, Tag } from '@/ui/components';

// type
import { FontWeight, Size } from '@/types';

// icons
import { ChatBox, File, Schedule } from '@/ui/components/Icons';
import { formatDate } from '@/utils';

export default async function PostDetail({ postId }: { postId: string }) {
  const postResponse = await getPostDataById(postId);
  const { data: listComments } = await getCommentByPostId(postId);

  const { data: postCategoriesDataResponse } = await getCategoryByPostId({
    postId,
  });
  const categoryId = postCategoriesDataResponse[0]?.attributes.categoryId ?? '1';
  console.log('categoryId', categoryId);
  const categoryIdToNumber = parseInt(categoryId, 10);
  const { data: categoryDataResponse } =
    await getCategoryById(categoryIdToNumber);
  const { name: categoryType } = categoryDataResponse.attributes;

  const attributes = postResponse?.data?.attributes;

  const title = attributes?.title ?? 'Title is not available';
  const content = attributes?.content ?? 'Content is not available';
  const imageUrl = attributes?.imageUrl ?? '';
  const publicDate = formatDate(attributes?.publicationDate) ?? '';
  const commentCount = listComments?.length ?? '0';

  return (
    <section className="flex flex-1">
      <article className="flex-1">
        <div className="mb-7.5 lg:bg-white-90 lg:p-7 xl:mb-32.5 xl:h-151 rounded-xl">
          <Typography
            tag="h1"
            additionalClasses="text-md font-medium sm:text-xxl sm:font-normal mb-7"
          >
            {title}
          </Typography>
          <div className="relative mx-auto w-89.5 h-49.5 sm:w-full sm:h-84 lg:h-95 xl:w-full xl:h-134.5 2xl:h-149.5 2xl:w-268">
            <Image
              fill
              objectFit="cover"
              className="rounded-xl"
              src={imageUrl}
              alt={title}
              placeholder="blur"
              blurDataURL={generateRGBDataURL(64, 64, 64)}
              sizes="(max-width: 603px) 358px,
              (min-width: 604px) and (max-width: 1535px) 100vw,
              (min-width: 1535px) 1072px"
            />
          </div>
        </div>
        <div className="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4">
          <Tag icon={<Schedule />} content={publicDate} />
          <Tag
            icon={<ChatBox />}
            content={`Comments: ${commentCount}`}
            additionalClass="hidden sm:flex"
          />
          <Tag
            icon={<File />}
            content={`Categories: ${categoryType}`}
            additionalClass="hidden sm:flex"
          />
          <Tag
            icon={<ChatBox />}
            content={`${commentCount}`}
            additionalClass="sm:hidden"
          />
          <Tag
            icon={<File />}
            content={`${categoryType}`}
            additionalClass="sm:hidden"
          />
        </div>
        <section>
          {/* content */}
          {content?.split('\n').map((line, index) => (
            <Typography
              key={index}
              tag="h2"
              textSize={Size.MD}
              weight={FontWeight.Normal}
              additionalClasses="text-dark-100"
            >
              {line}
            </Typography>
          ))}
        </section>
      </article>
    </section>
  );
}
