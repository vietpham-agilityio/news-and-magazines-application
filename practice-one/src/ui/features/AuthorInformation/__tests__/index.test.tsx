import { render, waitFor } from '@testing-library/react';
import { expect } from '@jest/globals';

// service
import { getAuthorById, getPostDataById } from '@/services';

// component
import { AuthorInformation } from '@/ui/features';

// mock data
import { mockAuthorData, mockPostDetailData } from '@/mocks';

jest.mock('../../../../services', () => ({
  getAuthorById: jest.fn(),
  getPostDataById: jest.fn(),
}));

describe('AuthorInformation component', () => {
  beforeEach(() => {
    (getAuthorById as jest.MockedFunction<typeof Object>).mockResolvedValue(
      mockAuthorData
    );
    (getPostDataById as jest.MockedFunction<typeof Object>).mockResolvedValue(
      mockPostDetailData
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should render match with snapshot.', async () => {
    const { container } = render(
      await AuthorInformation({
        postId: '1',
      })
    );

    waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
