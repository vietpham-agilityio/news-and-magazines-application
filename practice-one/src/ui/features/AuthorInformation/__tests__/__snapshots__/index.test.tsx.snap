// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorInformation component Should render match with snapshot. 1`] = `
<div>
  <div
    class="bg-white-90 flex p-[15px] mb-[25px] rounded-xl"
  >
    <div
      style="width: 87px; height: 87px; position: relative;"
    >
      <img
        alt="Avatar author"
        class="rounded-xl"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="87px"
        src="/_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fwww.unsplash.com%2Fimage%3Fv&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: cover; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      />
    </div>
    <div
      class="flex-1 ml-2.5 mt-3.5"
    >
      <div
        class="flex items-center justify-between"
      >
        <h2
          aria-level="2"
          class="text-sm font-medium text-dark-100"
          role="heading"
        >
          Benzamine
        </h2>
        <h2
          aria-level="2"
          class="text-xs font-light text-dark-75"
          role="heading"
        >
          27 posts
        </h2>
      </div>
      <div
        class="mt-2"
      >
        <button
          aria-label="button-element"
          class="flex items-center rounded-xl bg-gradient-to-bl from-sunSet-0 to-sunSet-100 py-2.5 pr-6 pl-4 undefined hover:opacity-80"
        >
          <i
            class="mr-2 undefined"
          >
            <svg
              fill="none"
              height="12"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.688 5.75c0 .492-.41.902-.876.902H6.876v3.938c0 .465-.41.848-.875.848a.858.858 0 0 1-.875-.848V6.652H1.187c-.492 0-.875-.41-.875-.902 0-.465.383-.848.875-.848h3.938V.965c0-.492.383-.902.875-.902.465 0 .875.41.875.902v3.937h3.938c.464-.027.874.383.874.848Z"
                fill="#FFF"
              />
            </svg>
          </i>
          <h2
            aria-level="2"
            class="text-xsm font-medium text-white-100 leading-5"
            role="heading"
          >
            Follow
          </h2>
        </button>
      </div>
    </div>
  </div>
</div>
`;
